contracts/poap/src/contract.rs

```diff
@@ -0,0 +1,1375 @@
+        assert_eq!(
+            ContractError::StartTimeBeforeCurrentTime {
+                current_time: env.block.time,
+                start_time: Timestamp::from_seconds(start)
+            },
+            init_result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn instantiate_with_event_start_equal_current_time_error() {
+        let mut deps = mock_dependencies();
+        let env = mock_env();
+        let info = mock_info(ADMIN, &vec![]);
+        let mut init_msg = get_valid_init_msg(1);
+
+        let start = env.block.time.nanos();
+        init_msg.event_info.start_time = Timestamp::from_nanos(start);
+        init_msg.event_info.end_time = Timestamp::from_nanos(start + 600);
+
+        let init_result = instantiate(deps.as_mut(), env, info, init_msg);
+        assert_eq!(
+            ContractError::StartTimeBeforeCurrentTime {
+                current_time: Timestamp::from_nanos(start),
+                start_time: Timestamp::from_nanos(start)
+            },
+            init_result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn instantiate_with_event_end_before_current_time_error() {
+        let mut deps = mock_dependencies();
+        let env = mock_env();
+        let info = mock_info(ADMIN, &vec![]);
+        let mut init_msg = get_valid_init_msg(1);
+
+        // Create a start time 200 seconds before the current block time
+        let start = env.block.time.seconds() - 200;
+        let end = env.block.time.seconds() - 100;
+        init_msg.event_info.start_time = Timestamp::from_seconds(start);
+        // Start time 100 seconds before the current block time
+        init_msg.event_info.end_time = Timestamp::from_seconds(end);
+
+        let init_result = instantiate(deps.as_mut(), env.clone(), info, init_msg);
+        assert_eq!(
+            ContractError::EndTimeBeforeCurrentTime {
+                current_time: env.block.time,
+                end_time: Timestamp::from_seconds(end)
+            },
+            init_result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn instantiate_with_event_end_equal_current_time_error() {
+        let mut deps = mock_dependencies();
+        let env = mock_env();
+        let info = mock_info(ADMIN, &vec![]);
+        let mut init_msg = get_valid_init_msg(1);
+
+        let start = env.block.time.seconds() - 200;
+        let end = env.block.time.nanos();
+        init_msg.event_info.start_time = Timestamp::from_seconds(start);
+        init_msg.event_info.end_time = Timestamp::from_nanos(end);
+
+        let init_result = instantiate(deps.as_mut(), env, info, init_msg);
+        assert_eq!(
+            ContractError::EndTimeBeforeCurrentTime {
+                current_time: Timestamp::from_nanos(end),
+                end_time: Timestamp::from_nanos(end),
+            },
+            init_result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn instantiate_with_event_start_after_end_error() {
```

Comment:
These test cases can be merged to be `instantiate_with_invalid_msg_error` with one case since they are well tested inside `msg.validate`:
1. instantiate_with_event_start_after_end_error
2. instantiate_with_event_start_equal_end_error
3. instantiate_with_invalid_poap_uri_error
4. instantiate_with_non_ipfs_poap_uri_error
5. instantiate_with_invalid_event_uri_error
6. instantiate_with_non_ipfs_event_uri_error

Reply:
This test was a single case before and then got splitted 😅
For consistence I will prefere to keep a single test for each case

Reply:
Okay, just thinking about they are covered by `msg.validate` already so we can reduce some tests in the `contract.rs`. I am fine with remaining them.

Reply:
I understand, but a test imo should ensure that a certain functionality behaves correctly regardless of how it is written internally. If we have just one test we are making an implicit assumption that the `validate` function will be called

Reply:
Yeah, you are right. Calling `msg.validate` is not the thing must to do in the contract, unlike chain app must run `msg.ValidateBasic` before executing it. I am mixing up them.